### ISSUE TITLE
blackbox stateful register when convert p4 Tofino to Bmv2

### ISSUE THREAD

**Issue by cseuk6** [p4lang/p4c#2686]
Hi, I want to running switchML program, which is originally written for Tofino backend. 
I have to convert the code written for the Tofino backend to codes written for Bmv2 backend. During this, I encountered a question, the blackbox_type stateful_alu in Tofino/stateful_alu_blackbox.p4 has no substitute in Bmv2, and the register_lo & register_hi are both set for Tofino hardware.  
So how can I substitute these structure in my program for Bmv2?

**Comment by hesingh** [comment]
You should contact Tofino support to help you get more details.  I think the p4c bmv2 supports register - at least the [JSON format ](https://github.com/p4lang/behavioral-model/blob/main/docs/JSON_format.md) does.

**Comment by mbudiu-vmw** [comment]
There is no way to do this with the open source tools as they exist today.
However, if you are interested to contribute, the open-source tools could be modified to support such extensions. This is non-trivial amount of work. You would have to modify bmv2 itself to support such primitives, then create a new architectural model that extends v1model.p4, and finally have the compiler generate code to support the new primitives in bmv2.

**Comment by jfingerh** [comment]
It should be straightforward to translate Tofino register access code into code that works on the v1model architecture for BMv2.  basically every stateful ALU is one or a pair of values of type `bit<W>`, and you would call `read` first, do whatever modifications are required, then `write` back the modified value.  For a pair of `bit<32>` values, for example, you can have a v1model register storing a `bit<64>` value, and use bit slices to separate it into 2 32-bit values, then put the modified values back together using P4_16 concatenation to create the modified `bit<64>` value to `write`.

**Comment by hesingh** [comment]
How did you use runtime-index into array on Tofino? The switchML algorithm uses runtime-index and there is no asic that supports runtime-index.  The bmv2 backend supports runtime-index.

**Comment by jfingerh** [comment]
This is going to sound like a strange question, given I work at Intel, but do folks have a link handy to the published switchml code you are looking at when asking these questions?  I'd rather answer questions about that specifically.

**Comment by jfingerh** [comment]
@hesingh I agree there might be few P4 implementations that support run-time index into header stacks, but supporting a run-time index when reading and writing P4 register arrays is supported by all P4 implementations I know of that support registers, just like they support run-time index for indirect (sometimes called indexed) counters and meters.

**Comment by ShailendraKK** [comment]
> This is going to sound like a strange question, given I work at Intel, but do folks have a link handy to the published switchml code you are looking at when asking these questions? I'd rather answer questions about that specifically.

@jfingerh, Here is the link to the section of the published code https://github.com/in-ATP/ATP/blob/main/p4src/includes/common.p4 that deals with blackbox stateful register

I think @cseuk6  is also referring to the same link. I am also facing the same issue while converting the Tofino p4 code to Netronome SmartNIC p4. Thanks

**Comment by jafingerhut** [comment]
I have assigned this issue to me, as I would like to use the opportunity to publish some examples of P4 programs using register arrays in the style of v1model and PSA (that use read() and write() methods) and programs with the same behavior that work for the Tofino Native Architecture.  Please be patient, as it may be a couple of weeks before I find the time to get them in publishable form.

**Comment by 122413693** [comment]
> > This is going to sound like a strange question, given I work at Intel, but do folks have a link handy to the published switchml code you are looking at when asking these questions? I'd rather answer questions about that specifically.
> 
> @jfingerh, Here is the link to the section of the published code https://github.com/in-ATP/ATP/blob/main/p4src/includes/common.p4 that deals with blackbox stateful register
> 
> I think @cseuk6 is also referring to the same link. I am also facing the same issue while converting the Tofino p4 code to Netronome SmartNIC p4. Thanks

Excuse me, I also find this question about blackbox stateful register run in the Bmv2. I just want to know wethere it has been solved?

**Comment by jafingerhut** [comment]
If your question is about the behavior of the blackbox stateful register as used in this repository's P4 code: https://github.com/in-ATP/ATP/blob/main/p4src/includes/common.p4

and how to rewrite it to run using the v1model architecture implemented by BMv2, I see at least three possible approaches you could take to finding the answer:

(a) Ask the authors of the code in that Github repository if they can help create a v1model version for you, or help you understand what the existing P4 code is doing so that you can write a v1model version of it.

(b) Learn enough about the blackbox stateful register constructs in the Tofino P4_14 compiler to understand what it is doing in general, and then you would understand how to write the equivalent v1model version of that code.  One way to do so would be to get a support contract with Intel to get official customer support on the Tofino P4_14 compiler, e.g. by following instructions here: https://forum.p4.org/t/intel-r-tofino-tm-family-tna-and-p4studio-questions/504

(c) Hope that someone reads this comment, and volunteers to teach you about the behavior of the Tofino P4_14 compiler blackbox stateful register features so that you can accomplish (a) or (b).  (I am not the person who will volunteer to do that, as first I would take at least several days to learn those P4_14 compiler features in detail myself before I could teach them, and I do not have any interest in spending my time that way.)

**Comment by jafingerhut** [comment]
Unassigning this issue from me, as I am unlikely to ever spend time publicly documenting the Tofino P4_14 compiler blackbox stateful language features in enough detail to be useful to anyone.
